pkg/virt-handler/container-disk/mount.go

```diff
@@ -367,3 +363,199 @@ func (m *mounter) ContainerDisksReady(vmi *v1.VirtualMachineInstance, notInitial
+				return err
+			}
+
+			file, err := os.Create(targetPath)
+			if err != nil {
+				return err
+			}
+			file.Close()
+
+			out, err := virt_chroot.MountChroot(sourcePath, targetPath, true).CombinedOutput()
+			if err != nil {
+				return fmt.Errorf("failed to bindmount %v: %v : %v", kernelBootName, string(out), err)
+			}
+
+			return nil
+		}
+
+		if err = mount(kb.InitrdPath, targetInitrdPath); err != nil {
+			return err
+		}
+		if err = mount(kb.KernelPath, targetKernelPath); err != nil {
+			return err
+		}
+	}
+
+	if verify {
+		mounted, err := areKernelArtifactsMounted(targetDir, targetInitrdPath, targetKernelPath)
+		if err != nil {
+			return fmt.Errorf("failed to check if kernel artifacts are mounted. error: %v", err)
+		} else if !mounted {
+			return fmt.Errorf("kernel artifacts verification failed")
+		}
+	}
+
+	return nil
+}
+
+func (m *mounter) UnmountKernelArtifacts(vmi *v1.VirtualMachineInstance) error {
+	if !util.IsKernelBootDefinedProperly(vmi) || vmi.UID == "" {
+		return nil
+	}
+
+	log.DefaultLogger().Object(vmi).Infof("unmounting kernel artifacts")
+
+	kb := vmi.Spec.Domain.Firmware.KernelBoot.Container
+
+	record, err := m.getMountTargetRecord(vmi)
+	if err != nil {
+		return fmt.Errorf("failed to get mount target record: %v", err)
+	} else if record == nil {
+		log.DefaultLogger().Object(vmi).Warning("Cannot find kernel-boot entries to unmount")
+		return nil
+	}
+
+	unmount := func(targetDir string, artifactPaths ...string) error {
+		for _, artifactPath := range artifactPaths {
+			if artifactPath == "" {
+				continue
+			}
+
+			targetPath := filepath.Join(targetDir, filepath.Base(artifactPath))
+			if mounted, err := isolation.NodeIsolationResult().IsMounted(targetPath); err != nil {
+				return fmt.Errorf("failed to check mount point for containerDisk %v: %v", targetPath, err)
+			} else if mounted {
+				log.DefaultLogger().Object(vmi).Infof("unmounting container disk at targetDir %s", targetPath)
+
+				out, err := virt_chroot.UmountChroot(targetPath).CombinedOutput()
+				if err != nil {
+					return fmt.Errorf("failed to unmount containerDisk %v: %v : %v", targetPath, string(out), err)
+				}
+			}
+		}
+		return nil
+	}
+
+	for idx, entry := range record.MountTargetEntries {
+		targetDir := entry.TargetFile
+		if !strings.Contains(targetDir, containerdisk.KernelBootName) {
+			continue
+		}
+		log.DefaultLogger().Object(vmi).Infof("unmounting kernel artifacts in path: %s", targetDir)
+
+		if err = unmount(targetDir, kb.InitrdPath, kb.KernelPath); err != nil {
+			// Not returning here since even if unmount wasn't successful it's better to keep
+			// cleaning the mounted files.
+			log.Log.Object(vmi).Reason(err).Error("unable to unmount kernel artifacts")
+		}
+
+		err = os.Remove(targetDir)
+		if err != nil {
+			log.DefaultLogger().Object(vmi).Infof("cannot delete dir %s. err: %v", targetDir, err)
+		}
+
+		removeSliceElement := func(s []vmiMountTargetEntry, idxToRemove int) []vmiMountTargetEntry {
```

Comment:
I have read that this is the fastest "go way" to remove an element from a slice.. is there a better way?